### ISSUE TITLE
`-principal` warning in standard flags?

### ISSUE THREAD

**Issue by mmottl** [ocaml/dune#1052]
I believe it used to be the case that adding the `-principal` flag significantly increased compilation times.  Having just tried it on a large project, it seems that the difference is rather benign (less than 10%).

Is there a reason not to add it to the standard development flags by default?  I think it would help make source code more resilient to future compiler changes that modify type inference algorithms.

**Comment by diml** [comment]
Why not make it the default in the compiler itself?

**Comment by mmottl** [comment]
That's a good point.  Guaranteeing future compatibility wrt. changes to the inference algorithm seems like a noble enough goal for a compiler default.  I haven't tried out how `-principal` interacts with other flags.  E.g. maybe it becomes slower unless `-short-paths` (dune dev default), is passed.

I guess it would be most effective if Jane Street could discuss this internally and, if necessary, propose this change to the compiler team.

**Comment by diml** [comment]
I defer this to @lpw25 and @trefis as I don't have much experience with this flag.

**Comment by trefis** [comment]
For the performance implications if I had to guess I'd say it might matter if you play with objects a lot (because there should be less sharing when compiling with `-principal`), but I don't have any idea of the scale, nor do I have a test case at hand.

Apart from that, not sure turning on `-principal` is such a good idea right now:
```
# module M = struct type t = A | B end;;
module M : sig type t = A | B end
# let f = function M.A -> true | B -> false;;
val f : M.t -> bool = <fun>
# #principal true;;
# let f = function M.A -> true | B -> false;;
Characters 31-32:
  let f = function M.A -> true | B -> false;;
                                 ^
Error: Unbound constructor B
```

**Comment by Alizter** [comment]
Not sure if this is something we really want to enforce in Dune. Users are welcome to use these flags in their own libraries.

**Comment by trefis** [comment]
FTR the behavior I was talking about before has been fixed (by ocaml/ocaml#1931) in 4.12:

```ocaml
# #principal true;;
# let f = function M.A -> true | B -> false;;
                                 ^
Warning 18 [not-principal]: this type-based constructor disambiguation is not principal.
val f : M.t -> bool = <fun>
```
